app/vmauth/main.go

```diff
@@ -160,7 +160,14 @@ func processRequest(w http.ResponseWriter, r *http.Request, ui *UserInfo) {
 	if up == nil {
 		missingRouteRequests.Inc()
 		if ui.DefaultURL == nil {
-			httpserver.Errorf(w, r, "missing route for %q", u.String())
+			// Authorization should be requested for http requests without credentials
+			// to a route that is not in the configuration for unauthorized user
+			if ui.BearerToken == "" && ui.Username == "" {
+				w.Header().Set("WWW-Authenticate", `Basic realm="Restricted"`)
+				http.Error(w, fmt.Sprintf("missing unauthorized route for %q", u.String()), http.StatusUnauthorized)
+				return
+			}
+			http.Error(w, fmt.Sprintf("missing route for %q", u.String()), http.StatusForbidden)
```

Comment:
this looks brittle...
It will only work if unathorized_user doesn't have route matching the request... And if it will have default_url - it will catch everything. 

I don't know how to do it better though. 
Let's describe the problem once again:
1. vmauth has two users
2. user `foo` with url_prefix: "http://my-address". The expectation is that any request to "http://my-address" will require BasicAuth. 
3. user `unathorized_user` for catching requests which failed BasicAuth

The problem here with p2 for users who use browser. Browser sends the first request without any auth. It expects to get "WWW-Authenticate" header in response if auth is needed. When it gets it - it displays the popup for user to enter creds.

Based on this, 1st request from browser will be always with no creds, and will always fall into unauthorized.

Ideal solution would be, probably, to always ask for "WWW-Authenticate" and then do the routing. The challenge here is to differentiate between those clients who received the header and provided auth, and those who received the header and failed to provide auth. Some tracking for already seen clients?

Reply:
I think you overthought :)

Let's try to imagine a scenario where the current approach would not work:
- The `users` section has a route for openning in the browser (i.e. it's static) and after authorization it will lead to some "backend" specifically for a particular one 
- The `unauthorized_user` section has the exact same route for openning in the browser, which should not ask for authorization and leads to a different backend.
- Some magical way when opening the route in a browser we need to understand whether to ask for authorization or not.

I don't think that scenario is likely to happen, but we can write in the docs about priority of `unathorized_user` for same route as in `users` if you open it in browser without credentials.

Reply:
Probably you're right.

> I don't think that scenario is likely to happen, but we can write in the docs about priority of unathorized_user for same route as in users if you open it in browser without credentials.

Would you mind adding the corresponding doc change?

Reply:
Done